### ISSUE TITLE
Refactor HomomorphicCommitment

### ISSUE THREAD

**Issue by philipr-za** [tari-project/tari#139]
The `HomomorphicCommitment` trait is going to be a heavily used component of Tari. A goal of the trait is make sure that we do not copy the base points (G and H) for every commitment made in the system. However, we need to support the use of different bases and cannot assume the use of only one pair of G and H.

The current approach of keeping track of a reference to the Base has turned out to be quite clunky and the design of this trait should be revisited to make it more ergonomic and modular.

**Comment by CjS77** [comment]
One approach could be to create a Factory function,

So we remove the `HomomorphicCommitment::new` entirely, but leave the reference to base in the struct implementations. Then we have a `HomomorphicCommitmentFactory` trait that olds a reference to the Base, and can implicitly assign the base. Something like

```
pub trait HomomorphicCommitment<K> {
    fn open(&self, k: &K, v: &K) -> bool;
    fn as_bytes(&self) -> &[u8];
}

pub trait HomomorphicCommitmentFactory<K> {
    type Base;
    type C: HomomorphicCommitment;

    fn new(base: &'static Self::Base) -> Self;
    fn create(k: &K, v: &K) -> Self::C;
}
```

I think we should also drop the dependency on Scalar so that we don't have to pull the dalek libraries through to the domain layer

**Comment by CjS77** [comment]
Fixed in #174
